docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,167 @@
+
+In this context, when the sender of a cross-chain transfer is **not** the source where the tokens
+were originated, the protocol prefixes the denomination with the port and channel identifiers in the
+following format:
+
+```typescript
+prefix + denom = {destPortN}/{destChannelN}/.../{destPort0}/{destChannel0}/denom
+```
+
+Example: transfering `100 uatom` from port `HubPort` and channel `HubChannel` on the Hub to
+Ethermint's port `EthermintPort` and channel `EthermintChannel` results in `10
+EthermintPort/EthermintChannel/uatom`, where `EthermintPort/EthermintChannel/uatom` is the new
+denomination on the receiving chain.
+
+In the case those tokens are transfered back to the Hub (i.e the **source** chain), the prefix is
+trimmed and the token denomination updated to the original one.
+
+### Problem
+
+The problem of adding additional information to the coin denomination is twofold:
+
+1. The ever increasing length if tokens are transfered to zones other than the source:
+
+If a token is transfered `n` times via IBC to a sink chain, the token denom will contain `n` pairs
+of prefixes, as shown on the format example above. This is supposes a problem because, while port
+and channel identifiers don't have a cap on the maximum lenght, the SDK `Coin` type only accepts
+denoms up to 64 characters.
+
+This can result undesired behaviours such as tokens not being abled to be transfered to multiple
+sink chains if the denomination exceeds the lenght or unexpected `panics` due to denomination
+validation on the receiving chain.
+
+2. The existence of special characters and upercase letters on the denomination:
+
+In the SDK everytime a Coin is initialized trhough the constructor function `NewCoin`, a validation
+of a coin's denom is performed according to a
+[Regex](https://github.com/cosmos/cosmos-sdk/blob/a940214a4923a3bf9a9161cd14bd3072299cd0c9/types/coin.go#L583),
+where only lowercase alphanumeric characters are accepted. While this is deriable for native denoms
+to keep a clean UX, it supposes a challenge for IBC as ports and channels maight be randomly
+generated with special carracters and uppercases as per the [ICS 024 - Host
+Requirements](https://github.com/cosmos/ics/tree/master/spec/ics-024-host-requirements#paths-identifiers-separators)
+specification.
+
+## Decision
+
+Introduce a new `Trace` field to the SDK's `Coin` type so that the two problems are mitigated.
+
+```proto
+// Coin defines a token with a denomination and an amount.
+//
+// NOTE: The amount field is an Int which implements the custom method
+// signatures required by gogoproto.
+message Coin {
+  option (gogoproto.equal) = true;
+
+  string denom  = 1;
+  string amount = 2 [(gogoproto.customtype) = "Int", (gogoproto.nullable) = false];
+  // trace the origin of the token. Every time a Coin is transferred to a chain that's not the souce
+  // of the token, a new item is inserted to the first position.
+  repeated Trace trace = 3;
```

Comment:
Thanks for the elaborate and well throughout ADR @fedekunze, but I can see that I'll already have concerns with this because we're conflating concerns and use-cases w/ IBC. Specifically, I do not believe `trace` should be a first-class citizen field of `Coin`. Please consider instead using some notion of metadata.

Reply:
I agree that we should use metadata and not add this to `Coin` itself. @fedekunze I have some ideas on how to address this more generally and will write something up.

Reply:
So I realize what I have in mind will be a larger ADR that also covers supply. Hoping to get to that soon as these are things I've thought about for some of Regen's use case which allows for some modules to arbitrarily create approved denom's.

Anyway the gist of it for this use case is that from the perspective of the bank module, IBC transfer is just minting or burning coins for some denom which it has the right to manage.

I propose we use prefixes for these denoms such as `ibc:` or `ibc/` and then IBC assigns an auto-generated ID that looks something like `ibc:2g8sd6h`. Or maybe `ibc:atom:2g8sd6h` to be more descriptive if the source denom is `atom`. Then the `Trace` information is something that gets managed as metadata by the ibc transfer module. So the ibc transfer keeper would then have a method like `DescribeIBCDenom(denom string) IBCDenomMetadata` to retrieve the `Trace` data.

One problem that I pointed out in the call is that balances are indexed by `denom`. So adding `Trace` would break that in addition polluting any balances with all of that metadata. Instead we can just treat `denom`s as dumb identifiers which can be fully described by their metadata.

To take it a step further, I would restrict which modules can mint/burn which denoms. So `x/mint` would get rights to `atom`, `x/ibc-transfer` can mange the whole `ibc:*` prefix, etc. But that gets into supply which I'd love to address too, but later after stargate.

Reply:
This makes sense, although for now (IBC 1.0) I think we should do the simplest thing which solves our main problem, and hash the denominations in the `20-transfer` module to allow for arbitrary port/channel paths, and keep a lookup table so that we can lookup the port(s) & channel(s) from the denomination.